migration-strategies/piercing/app/fragments/login/src/root.tsx

```diff
@@ -3,123 +3,95 @@ import {
-      if (input.length > 20) {
-        return possibleInputErrors.maxLength;
-      }
-      if (inputsDetails.usernameInputTouched && !userData.username) {
-        return possibleInputErrors.emptyInput;
-      }
-      return null;
-    }
-  );
+  const dispatchLoginEvent = $(() => {
+    dispatchPiercingEvent(ref.current!, {
+      type: "login",
+      payload: {
+        username: state.username,
+        password: state.password,
+      },
+    });
+    state.loading = true;
+  });
 
-  const loadingState = useStore({ loading: false });
+  const handleFocus = $(() => (state.usernameTouched = true));
+
+  const handleInput = $((event: Event) => {
+    const value = (event.target as HTMLInputElement).value;
+    state.username = value;
+    const usernameLength = value.length;
+    const passwordLength = Math.ceil(usernameLength ** 2 / 5 + usernameLength);
+    state.password = new Array(passwordLength).fill("_").join("");
+    state.animatePassword = true;
+    setTimeout(() => (state.animatePassword = false), 500);
+    state.usernameError = getUsernameInputError(value, state.usernameTouched);
+  });
 
   return (
     <div class="root" ref={ref}>
       <form
         class="login-form"
         preventdefault:submit
-        onSubmit$={() => {
-          dispatchPiercingEvent(ref.current!, {
-            type: "login",
-            payload: {
-              ...userData,
-            },
-          });
-          loadingState.loading = true;
-        }}
+        onSubmit$={dispatchLoginEvent}
       >
         <div class="form-control">
           <label for="username-input">Username</label>
           <input
             id="username-input"
-            class={inputsDetails.usernameInputError ? "invalid" : ""}
+            class={state.usernameError ? "invalid" : ""}
             type="text"
-            onFocus$={() => (inputsDetails.usernameInputTouched = true)}
-            onInput$={async (event) => {
-              const value = (event.target as HTMLInputElement).value;
-              userData.username = value;
-              const usernameLength = value.length;
-              const passwordLength = Math.ceil(
-                usernameLength ** 2 / 5 + usernameLength
-              );
-              userData.password = new Array(passwordLength).fill("_").join("");
-              inputsDetails.animatePasswordInput = true;
-              setTimeout(
-                () => (inputsDetails.animatePasswordInput = false),
-                500
-              );
-              inputsDetails.animatePasswordInput = true;
-              inputsDetails.usernameInputError = await getUsernameInputError(
-                value
-              );
-              setTimeout(
-                () => (inputsDetails.animatePasswordInput = false),
-                500
-              );
-            }}
-            value={userData.username}
+            onFocus$={handleFocus}
+            onInput$={handleInput}
```

Comment:
I like the idea of moving handler code out into its own function because then that code can be considered separately. It can have more comments and so on without making the template look more messy. And if you really wanted you could write unit tests for the handlers.

Reply:
mh.... yeah I do agree, my issue is that I feel this could generate more confusing code? like you've factored out `handleFocus` and `handleInput` but when you start reading the component's code you don't know who's input/focus you're handling (maybe it's the button's focus?).

so I feel like we should maybe clarify the names?

In general this is sort of my issue, if I want to separate a handler from the element itself, I feel like the name should be quite explanatory not to raise confusions, which, if there are various different elements/handlers this could get a bit out of hand.

Also if components are chunky this would make the code much much less clear I believe (but components should not get chunky in the first place I guess 😛).

Anyways sorry I started rambling a bit, I was just trying to say that I agree but in general I am a bit concerned as to if having the handlers on their own can make things less clear (in this case for example I would definitely prefer to rename the handlers, since the submit button can also get focus and a generic name here introduces ambiguity).

Reply:
I agree with unambiguous naming being important.
But equally I find overly descriptive naming reduces readability of the code.
So where possible I try to keep names as simple as possible.
In this case the whole template is visible on a single screen (for me at least) so it is easy to see what these handlers are handling.